qiskit/algorithms/linear_solvers/observables/matrix_functional.py

```diff
@@ -0,0 +1,120 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""The matrix functional of the vector solution to the linear systems."""
+from typing import Union, Optional, List
+import numpy as np
+from scipy.sparse import diags
+
+from qiskit.aqua.operators import PauliSumOp
+from qiskit import QuantumCircuit
+from qiskit.opflow import I, Z, TensoredOp
+from .linear_system_observable import LinearSystemObservable
+
+
+class MatrixFunctional(LinearSystemObservable):
+    """A class for the matrix functional of the vector solution to the linear systems."""
+
+    def __init__(self, main_diag: float, off_diag: int) -> None:
+        """
+        Args:
+            main_diag: The main diagonal of the tridiagonal Toeplitz symmetric matrix to compute
+             the functional.
+            off_diag: The off diagonal of the tridiagonal Toeplitz symmetric matrix to compute
+             the functional.
+        """
+        self._main_diag = main_diag
+        self._off_diag = off_diag
+
+    def observable(self, num_qubits: int) -> Union[PauliSumOp, List[PauliSumOp]]:
```

Comment:
Like in an another version of an implementation of this just checking if the return type matches since I see its an expression with TensoredOp in there. Maybe it ends up as PauliSumOp in the case due to the expression

Reply:
changed